File: src/CSharp/badgeup/badge/ProlificLEDBadge.cs
namespace badgeup.badge
{
    using System.IO.Ports;
    using System.Text;
    using System;
    using System.IO;

    /// <summary>
    /// Class to encapsulate Properties and Methods needed to communicate with the Prolific LED
    /// Badge over COM
    /// </summary>
    public class ProlificLEDBadge : COMBadge
    {
        public const byte StartByte        = 0x00;          //
        public const byte PaddingByte      = 0x00;          //
        public const byte ControlByteOne   = 0x02;          //
        public const byte ControlByteTwo   = 0x31;          //
        public const byte ControlByteThree = 0x33;          //
        public const uint CheckSumMod      = 0x100;         //256 in hex used get a 1-byte result

        public const int MaxLength         = 250;           //Maximum message length

        /// <summary>
        /// Default Constructor - Sets the Prolific LED Badge specific COM Port Settings observed from testing/portmon
        /// </summary>
        public ProlificLEDBadge() 
        {
            this.BaudRate = 38400;
            this.Parity = Parity.None;
            this.WordSize = 8;
            this.Stop = StopBits.One;
            this.QueryBadgeValue = "Prolific";
        }
        /// <summary>
        /// Initializes and prepares the Prolific LED Badge for
        /// communication.
        /// </summary>
        public override void Open()
        {
            string badgePort;
            if (this.TryGetBadgePort(out badgePort))
            {
                this.ComPort = badgePort;
            }
            else
            {
                throw new COMBadgeNotFoundException();
            }
        }
        /// <summary>
        /// Close the connection to the Prolific Badge
        /// </summary>
        public override void Close()
        {
            //do nothing
        }
        /// <summary>
        /// Sets an message on the badge
        /// </summary>
        /// <param name="message">An LEDMessage containing the message to set</param>
        public void SetMessage(LEDMessage message)
        {
            this.SetMessages(new LEDMessage[] { message });
        }
        /// <summary>
        /// Sets the messages on the badge
        /// </summary>
        /// <param name="messages">An Array of up to 6 messages to set</param>
        public void SetMessages(LEDMessage[] messages)
        {
            byte[] badgeBytes = null;
            using (MemoryStream bytePattern = new MemoryStream())
            {
                //write the start byte
                bytePattern.WriteByte(StartByte);

                //get number of messages to process
                int numMsgs = 6;
                if (messages.Length < numMsgs)
                    numMsgs = messages.Length;

                //messages enabled flag
                uint msgCount = 0x00;

                //get the byte pattern foreach message
                for (int m = 0; m < numMsgs; m++)
                {
                    byte[] badgeMsg = CreateMessageBytePattern(messages[m].Message, m, messages[m].Style, messages[m].Speed);
                    bytePattern.Write(badgeMsg, 0, badgeMsg.Length);

                    msgCount += msgCount + 0x01;
                }

                //write end byte pattern
                bytePattern.WriteByte(ControlByteOne);
                bytePattern.WriteByte(ControlByteThree);
                bytePattern.WriteByte((byte)msgCount);

                //get final byte pattern
                badgeBytes = bytePattern.ToArray();
            }

            //write bytes to the badge
            this.WriteBytes(badgeBytes);
        }
        /// <summary>
        /// Clears all the messages from the badge
        /// </summary>
        public void ClearMessages()
        {
            //Sets the messages flag to 0000 0000 = no messages enabled
            this.EnableMessages(0x00);
        }
        /// <summary>
        /// Enables messages on the badge based on a message pattern
        /// 0x00 = No messages
        /// 0xFF = All messages 6 character string + 2 image
        /// </summary>
        /// <param name="messagePattern"></param>
        public void EnableMessages(byte messagePattern)
        {
            byte[] enablePattern = new byte[] { 0x00, ControlByteOne, ControlByteThree, messagePattern };
            this.WriteBytes(enablePattern);
        }
        /// <summary>
        /// Creates the message byte pattern expected by the Prolific LED Badge
        /// </summary>
        /// <param name="msg">A string containing the message, must be less/== 250 chars</param>
        /// <param name="msgId">An int identifying which message this is</param>
        /// <param name="msgStyle">A MessageStyle value (hold, snowing, etc)</param>
        /// <param name="msgSpeed">A MessageSpeed (one through five)</param>
        /// <returns>Message byte array</returns>
        private byte[] CreateMessageBytePattern(string msg, int msgId, MessageStyle msgStyle, MessageSpeed msgSpeed)
        {
            //Get the message offsets to use
            MessageOffsetFirst offset1  = MessageOffsetFirst.One;
            MessageOffsetSecond offset2 = MessageOffsetSecond.One;
            switch(msgId)
            {
                case 0:
                    offset1 = MessageOffsetFirst.One;
                    offset2 = MessageOffsetSecond.One;
                    break;
                case 1:
                    offset1 = MessageOffsetFirst.Two;
                    offset2 = MessageOffsetSecond.Two;
                    break;
                case 2:
                    offset1 = MessageOffsetFirst.Three;
                    offset2 = MessageOffsetSecond.Three;
                    break;
                case 3:
                    offset1 = MessageOffsetFirst.Four;
                    offset2 = MessageOffsetSecond.Four;
                    break;
                case 4:
                    offset1 = MessageOffsetFirst.Five;
                    offset2 = MessageOffsetSecond.Five;
                    break;
                case 5:
                    offset1 = MessageOffsetFirst.Six;
                    offset2 = MessageOffsetSecond.Six;
                    break;
            }

            //ensure length is within bounds
            if(msg.Length > MaxLength)
                msg = msg.Substring(0, MaxLength);

            //Assuming English and using ASCIIEncoding
            byte[] msgBytes = ASCIIEncoding.ASCII.GetBytes(msg);
            byte msgLength  = Convert.ToByte(msgBytes.Length);

            //Initialize data and looping counters
            byte[] msgData;
            int msgBytesRead = 0;
            byte messageBatch = 0x00;
            using (MemoryStream ms = new MemoryStream())
            {
                //Build the 4 message groups
                for (int group = 0; group < 4; group++)
                {
                    //checksum for each group
                    //calculated by: (control bits - 0x02 + data) % 0x100
                    uint checksum = 0;

                    switch (group)
                    {
                        case 0:

                            ms.WriteByte(ControlByteOne);
                            ms.WriteByte(ControlByteTwo);
                            ms.WriteByte((byte)offset1);
                            ms.WriteByte(messageBatch);
                            ms.WriteByte((byte)msgSpeed);
                            ms.WriteByte((byte)offset2);
                            ms.WriteByte((byte)msgStyle);
                            ms.WriteByte(msgLength);
                            checksum += (uint)(ControlByteTwo + (byte)offset1 + messageBatch + (byte)msgSpeed + (byte)offset2 + (byte)msgStyle + msgLength);
                            //write data for this message - only 60 chars/bytes because of control bytes
                            for (int m = 0; m < 60; m++)
                            {
                                //write message data
                                if (msgBytesRead < msgBytes.Length)
                                {
                                    byte current = msgBytes[msgBytesRead];
                                    ms.WriteByte(current);
                                    checksum += (uint)current;

                                    msgBytesRead++;
                                }
                                else
                                {
                                    ms.WriteByte(PaddingByte);
                                }
                            }
                            ms.WriteByte((byte)(checksum % 256));
                            break;
                        case 1:
                        case 2:
                            ms.WriteByte(ControlByteOne);
                            ms.WriteByte(ControlByteTwo);
                            ms.WriteByte((byte)offset1);
                            ms.WriteByte(messageBatch);
                            checksum += (uint)(ControlByteTwo + (byte)offset1 + messageBatch);
                            //write data for this message - 64 chars/bytes since no extra control bits
                            for (int m = 0; m < 64; m++)
                            {
                                //write message data
                                if (msgBytesRead < msgBytes.Length)
                                {
                                    byte current = msgBytes[msgBytesRead];
                                    ms.WriteByte(current);
                                    checksum += (uint)current;

                                    msgBytesRead++;
                                }
                                else
                                {
                                    ms.WriteByte(PaddingByte);
                                }
                            }
                            ms.WriteByte((byte)(checksum % 256));
                            break;
                        case 3:
                            ms.WriteByte(ControlByteOne);
                            ms.WriteByte(ControlByteTwo);
                            ms.WriteByte((byte)offset1);
                            ms.WriteByte(messageBatch);
                            checksum += (uint)(ControlByteTwo + (byte)offset1 + messageBatch);
                            //write data for this message - 62 chars/bytes
                            for (int m = 0; m < 62; m++)
                            {
                                //write message data
                                if (msgBytesRead < msgBytes.Length)
                                {
                                    byte current = msgBytes[msgBytesRead];
                                    ms.WriteByte(current);
                                    checksum += (uint)current;

                                    msgBytesRead++;
                                }
                                else
                                {
                                    ms.WriteByte(PaddingByte);
                                }
                            }
                            ms.WriteByte(PaddingByte);
                            ms.WriteByte(PaddingByte);
                            ms.WriteByte((byte)(checksum % 256));
                            break;
                    }
                    messageBatch += 0x40;
                }

                msgData = ms.ToArray();
            }

            return msgData;
        }
        /// <summary>
        /// LED Badge Message Scrolling/Update speed
        /// One is Slowest, Five is Fastest
        /// </summary>
        public enum MessageSpeed
        {
            One = 0x31,
            Two = 0x32,
            Three = 0x33,
            Four = 0x34,
            Five = 0x35
        }
        /// <summary>
        /// LED Badge Message Styles
        /// </summary>
        public enum MessageStyle
        {
            Hold = 0x41,
            Scrolling = 0x42,
            RainDown = 0x43,
            Flash = 0x44
        }
        /// <summary>
        /// LED Badge first message offset used to set a numbered message
        /// MessageOffsetFirst.One = first message's first offset
        /// </summary>
        public enum MessageOffsetFirst
        {
            One = 0x06,
            Two = 0x07,
            Three = 0x08,
            Four = 0x09,
            Five = 0x0A,
            Six = 0x0B
        }
        /// <summary>
        /// LED Badge second message offset used to set a numbered message
        /// MessageOffsetSecond.One = first message's second offset
        /// </summary>
        public enum MessageOffsetSecond
        {
            One = 0x31,
            Two = 0x32,
            Three = 0x33,
            Four = 0x34,
            Five = 0x35,
            Six = 0x36
        }
        /// <summary>
        /// LEDMessage - Encapsulates an LED Message for the Prolific LED Badge
        /// </summary>
        public class LEDMessage
        {
            public string Message {get; set;}       //message text
            public MessageStyle Style {get; set;}   //message style - hold, snow, scrolling
            public MessageSpeed Speed {get; set;}   //message speed - one through five (one is slow)

            //Defaults
            public static int MAXLENGTH = 250;  
            public static MessageStyle DEFAULTSTYLE = MessageStyle.Scrolling;
            public static MessageSpeed DEFAULTSPEED = MessageSpeed.One;

            public LEDMessage(string message)
            {
                //only allow messages of MAXLENGTH
                if(message.Length > MAXLENGTH)
                    message = message.Substring(0, (MAXLENGTH - 1));

                this.Message = message;
                this.Style = DEFAULTSTYLE;
                this.Speed = DEFAULTSPEED;
            }
        }
    }
}

File: src/CPy27/badgeup/badgeup/COMBadge/ProlificLEDBadge.py
import serial
import array

#import COMBadge base class functionality
from COMBadge import COMBadge
from COMBadge import COMBadgeNotFoundException

class ProlificLEDBadge(COMBadge):
    
    #constants
    StartByte        = 0x00
    PaddingByte      = 0x00
    ControlByteOne   = 0x02
    ControlByteTwo   = 0x31
    ControlByteThree = 0x33
    CheckSumMod      = 0x100        #256 in hex used get a 1-byte result

    MaxLength        = 250          #Maximum message length
    MaxMessages      = 6
    
    def __init__(self):
        super(COMBadge,self).__init__()
        
        self.BaudRate           = 38400
        self.Parity             = serial.PARITY_NONE
        self.WordSize           = 8
        self.Stop               = serial.STOPBITS_ONE
        self.QueryBadgeValue    = "Prolific"
    
    def Open(self):
        """Opens the Prolific LED Badge or throws an error"""
        badgePort = self.GetBadgePort()
        if(badgePort):
            self.ComPort = badgePort
        else:
            raise COMBadgeNotFoundException
    
    def Close(self):
        """ do nothing """

    def ClearMessages(self):
        """Clears all the messages from the badge - Sets the messages flag to 0000 0000 = no messages enabled"""
        self.EnableMessages(0x00)

    def EnableMessages(self, messagePattern):
        """ Enables messages on the badge based on a message pattern
            0x00 = No messages
            0xFF = All messages 6 character string + 2 image"""
        bytesToWrite = array.array('B')
        bytesToWrite.append(0x00)
        bytesToWrite.append(self.ControlByteOne)
        bytesToWrite.append(self.ControlByteThree)
        bytesToWrite.append(messagePattern)
        
        self.WriteBytes(bytesToWrite)

    def SetMessage(self, message):
        """Sets a message on the badge"""
        self.SetMessages([ message ])
        
    def SetMessages(self, messages):
        """Sets messages on the badge - up to six text messages"""
        
        #begin byte buffer and add starting bytes
        badgeBytes = bytearray()
        badgeBytes.append(self.StartByte)
        
        #get number of messages
        numMsgs = self.MaxMessages
        messagesLength = len(messages)
        if messagesLength < numMsgs :
            numMsgs = messagesLength
        
        #messages enabled flag
        msgCount = 0x00
        
        #for each message, write out the byte pattern
        for i in range(len(messages)):
            badgeBytes.extend(self.__CreateMessageBytePattern(messages[i].Message, i, messages[i].Style, messages[i].Speed))
            msgCount += 1
        
        #write end byte pattern
        badgeBytes.append(self.ControlByteOne)
        badgeBytes.append(self.ControlByteThree)
        badgeBytes.append(msgCount)
        
        for c in badgeBytes:
            print "'0x%X" % c        
        
        #write bytes to badge
        self.WriteBytes(badgeBytes)

    def __CreateMessageBytePattern(self, message, messageId, style, speed):
        """Creates the message byte pattern expected by the Prolific LED Badge"""
        
        #get message offsets
        offset1 = None
        offset2 = None
        if   messageId == 0:
            offset1 = MessageOffsetFirst.One
            offset2 = MessageOffsetSecond.One
        elif messageId == 1:
            offset1 = MessageOffsetFirst.Two
            offset2 = MessageOffsetSecond.Two
        elif messageId == 2:
            offset1 = MessageOffsetFirst.Three
            offset2 = MessageOffsetSecond.Three
        elif messageId == 3:
            offset1 = MessageOffsetFirst.Four
            offset2 = MessageOffsetSecond.Four
        elif messageId == 4:
            offset1 = MessageOffsetFirst.Five
            offset2 = MessageOffsetSecond.Five
        elif messageId == 5:
            offset1 = MessageOffsetFirst.Six
            offset2 = MessageOffsetSecond.Six

        #ensure length is within bounds
        if len(message) > self.MaxLength:
            message = message[1:self.MaxLength]
        
        #initialize buffer and looping counters
        messageBytes = array.array('B')
        msgBytesRead = 0
        messageBatch = 0x00
        
        for m in range(1, 5):
            checksum = 0x00
            
            if m == 1:
                messageBytes.append(self.ControlByteOne)
                messageBytes.append(self.ControlByteTwo)
                messageBytes.append(offset1)
                messageBytes.append(messageBatch)
                messageBytes.append(speed)
                messageBytes.append(offset2)
                messageBytes.append(style)
                messageBytes.append(len(message))
                checksum += self.ControlByteTwo + offset1 + messageBatch + speed + offset2 + style + len(message)
                
                for i in range(1, 61):
                    #write message data
                    if msgBytesRead < len(message):
                        current = ord(message[msgBytesRead])
                        messageBytes.append(current)
                        checksum += current
                        msgBytesRead += 1
                    else:
                        messageBytes.append(self.PaddingByte)
                
                messageBytes.append(checksum % 256)
                
            elif m == 2 or m == 3:
                messageBytes.append(self.ControlByteOne)
                messageBytes.append(self.ControlByteTwo)
                messageBytes.append(offset1)
                messageBytes.append(messageBatch)
                checksum += self.ControlByteTwo + offset1 + messageBatch
                
                for i in range (1, 65):
                    #write message data
                    if msgBytesRead < len(message):
                        current = ord(message[msgBytesRead])
                        messageBytes.append(current)
                        checksum += current
                        msgBytesRead += 1
                    else:
                        messageBytes.append(self.PaddingByte)
                
                messageBytes.append(checksum % 256)
            
            elif m == 4:
                messageBytes.append(self.ControlByteOne)
                messageBytes.append(self.ControlByteTwo)
                messageBytes.append(offset1)
                messageBytes.append(messageBatch)
                checksum += self.ControlByteTwo + offset1 + messageBatch
                
                for i in range (1, 63):
                    #write message data
                    if msgBytesRead < len(message):
                        current = ord(message[msgBytesRead])
                        messageBytes.append(current)
                        checksum += current
                        msgBytesRead += 1
                    else:
                        messageBytes.append(self.PaddingByte)
                
                messageBytes.append(self.PaddingByte)
                messageBytes.append(self.PaddingByte)
                messageBytes.append(checksum % 256)
            
            #increment message batch
            messageBatch += 0x40
        
        return messageBytes
    
#Helpers

class MessageSpeed(object):
    """LED Badge Message Scrolling/Update speed. One if slowest, Five is fastest"""
    One     = 0x31
    Two     = 0x32
    Three   = 0x33
    Four    = 0x34
    Five    = 0x35

class MessageStyle(object):
    """LED Badge Message Styles"""
    Hold        = 0x41
    Scrolling   = 0x42
    RainDown    = 0x43
    Flash       = 0x44

class MessageOffsetFirst(object):
    """LED Badge first message offset used to set a numbered message - MessageOffsetFirst.One = first message's first offset"""
    One     = 0x06
    Two     = 0x07
    Three   = 0x08
    Four    = 0x09
    Five    = 0x0A
    Six     = 0x0B
    
class MessageOffsetSecond(object):
    """LED Badge second message offset used to set a numbered message - MessageOffsetSecond.One = first message's second offset"""
    One     = 0x31
    Two     = 0x32
    Three   = 0x33
    Four    = 0x34
    Five    = 0x35
    Six     = 0x36

class LEDMessage(object):
    """LEDMessage - Encapsulates an LED Message for the Prolific LED Badge"""
    
    #constants
    MAXLENGTH = 250
    DEFAULTSTYLE = MessageStyle.Scrolling
    DEFAULTSPEED = MessageSpeed.One

    def __init__(self, message):
        self.Message    = message
        self.Style      = self.DEFAULTSTYLE
        self.Speed      = self.DEFAULTSPEED

        
        